teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,833 @@
+    contain, and we could validate at creation time that those keys exist
+  * For invalid uses that can only be caught at runtime (e.g. after fetching the task spec), the PipelineRun would fail
+
+### Risks and Mitigations
+
+[See TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477) for risks and mitigations
+specific to expanding ArrayOrString to support more types, including size limitations caused by our implementation
+of results ([pipelines#4012](https://github.com/tektoncd/pipeline/issues/4012)).
+
+### User Experience
+
+* Using the json schema syntax to describe the structure of the object is a bit verbose and might not be intuitive
+  to users; see [Alternatives](#alternatives) for some ways we might improve this.
+
+## Design Details
+
+### Declaring object results and params
+
+We would use [JSON object schema syntax](https://json-schema.org/understanding-json-schema/reference/object.html) to
+declare the structure of the object.
+
+Params example:
+
+```yaml
+  params:
+    - name: pull_remote
+      description: json schema has no "description" fields, so we'd have to include documentation about the structure in this field
+      schema:
+        type: object # see question below
+        properties:
+          url: {
+            type: string 
+          }
+          path: {
+            type: string
+          }
+```
+
+Results example:
+
+```yaml
+  results:
+    - name: sdist
+      description: json schema has no "description" fields, so we'd have to include documentation about the structure in this field
+      schema:
+        type: object
+        properties:
+          sha: {
+            type: string
+          }
+          path: {
+            type: string
+          }
+```
+
+#### What to do with the `type` field?
+
+Do we want to keep using [the existing `type` field](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#specifying-parameters)?
+i.e. do we use the top level `type: object` syntax, similar to how we declare `type: string` and `type: array` today, or
+do we want to move to always using `schema` and using the same syntax for arrays and strings as well (as suggested in
+[pipelines#1393](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-558833526))?
+
+For example, if we add the syntax suggested here to the existing `type` syntax, their uses would look like:
+
+```yaml
+params:
+- name: someURL
+  type: string
+- name: flags
+  type: array
+- name: sdist
+  schema:
+    type: object
+    properties:
+      sha: {
+        type: string
+      }
+      path: {
+        type: string
+      }
+```
+
+If we deprecated the type syntax and only used json schema, the above would become:
+
+```yaml
+params:
+- name: someURL
+  schema:
+    type: string
+- name: flags
+  schema:
+    type: array
+- name: sdist
+  schema:
+    type: object
+    properties:
+      sha: {
```

Comment:
I'm not quite sure I understand how this would work once we did start nesting things (if that's the plan). This type is will presumably be mapped into the actual go struct schema definition as something like:

```golang
type Property struct {
  Type string
}

type Properties []Property
```

But if we want to support nesting, how would we embed the jsonschema definition? Would it be something like:

```golang
type Properties struct {
  JSONSchema string
}
```